File: 0001-0500/0347. Top K Frequent Elements.java
// Medium
// https://leetcode.cn/problems/top-k-frequent-elements/

import java.util.*;

class Solution {
    // Use HashMap to count frequencies and minHeap to track top-k frequent elements
    // 1. Count the frequency of each element using a HashMap
    // 2. Use a min-heap (PriorityQueue) to keep the k most frequent elements
    // 3. If the heap size exceeds k, remove the least frequent element
    // 4. Extract elements from the heap to form the result
    // Time:O(nlogk),Space:O(N+K)
    // 适用于k相对小，n较大的时候
    public int[] topKFrequent(int[] nums, int k) {
        // use map to record the freq of elements <num, freq>
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());
        // put each key-val pairs to pq and check if size > k
        for (Map.Entry<Integer, Integer> e : map.entrySet()) {
            pq.offer(e);
            if (pq.size() > k) {
                pq.poll();
            }
        }
        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = pq.poll().getKey();
        }
        return res;
    }
}

class Solution2 {
    // sort+get topk
    // Time:O(nlogn),Space:O(n)
    // 适用于k比较接近n时，如果k比较小，其他方法更有效
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(map.entrySet());
        list.sort((a, b) -> b.getValue() - a.getValue());

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = list.get(i).getKey();
        }
        return res;
    }
}

class Solution3 {
    // bucketSort: build bucket, size is n+1. Each pos store the element has same Freq
    // iterate the bucket from high freq to low freq, collect topk
    // Time:O(n),Space:O(n)
    // 适用于k接近n、以及频率比较集中的情况
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        @SuppressWarnings("unchecked") // just remove warning
        List<Integer>[] bucket = new List[nums.length + 1];
        for (int i = 0; i < bucket.length; i++) {
            bucket[i] = new ArrayList<>();
        }
        // put the element to corresponding bucket
        for (Map.Entry<Integer, Integer> e : map.entrySet()) {
            int freq = e.getValue();
            bucket[freq].add(e.getKey());
        }
        // collect topk from high freq to low freq
        List<Integer> res = new ArrayList<>();
        for (int i = bucket.length - 1; i >= 0 && res.size() < k; i--) {
            for (int num : bucket[i]) {
                res.add(num);
                if (res.size() == k) break;
            }
        }
        // to array
        int[] ans = new int[k];
        for (int i = 0; i < k; i++) {
            ans[i] = res.get(i);
        }
        return ans;
    }
}

class Solution4 {
    // QuickSelect: map to record the freq, and transfer it to an arr
    // each element in arr is element-freq pair, and use quickselect to get topk
    // Time:O(n), the worse is O(n^2), Space:O(n)
    // 适用于k远小于n的时候
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        // put the freq to arr
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(map.entrySet());
        // use quick select to get topk
        quickSelect(list, 0, list.size() - 1, k);

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = list.get(i).getKey();
        }
        return res;
    }
    private void quickSelect(List<Map.Entry<Integer, Integer>> list, int left, int right, int k) {
        if (left >= right) return;
        int pivotIndex = partition(list, left, right);
        if (pivotIndex == k) return;
        else if (pivotIndex < k) quickSelect(list, pivotIndex + 1, right, k);
        else quickSelect(list, left, pivotIndex - 1, k);
    }
    private int partition(List<Map.Entry<Integer, Integer>> list, int left, int right) {
        Map.Entry<Integer, Integer> pivot = list.get(right);
        int pivotValue = pivot.getValue();
        int i = left;
        for (int j = left; j < right; j++) {
            if (list.get(j).getValue() > pivotValue) {
                Collections.swap(list, i, j);
                i++;
            }
        }
        Collections.swap(list, i, right);
        return i;
    }
}

File: 0001-0500/0347. Top K Frequent Elements.py
import heapq
from typing import Counter, List

# When k is small, and n is bigger
# Time:O(nlog k)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)

        heap = []
        for num, freq in count.items():
            heapq.heappush(heap, (freq, num))
            if len(heap) > k:
                heapq.heappop(heap)

        return [num for _, num in heap]

# when k is close to n
# Time: O(nlog n)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)

        sorted_nums = sorted(count.items(), key=lambda x: -x[1])

        return [num for num, _ in sorted_nums[:k]]

# when k is close to n and the freqs are relatively concentrated
# Time: O(n) 
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)
        n = len(nums)

        buckets = [[] for _ in range(n + 1)]
        for num, freq in count.items():
            buckets[freq].append(num)

        res = []
        for freq in range(n, -1, -1):
            for num in buckets[freq]:
                res.append(num)
                if len(res) == k:
                    return res

        return res

# when k is far small than n
# Time: O(n)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)
        items = list(count.items())

        def partition(left, right):
            pivot_freq = items[right][1]
            i = left
            for j in range(left, right):
                if items[j][1] > pivot_freq:
                    items[i], items[j] = items[j], items[i]
                    i += 1
            items[i], items[right] = items[right], items[i]
            return i

        def quick_select(left, right, k):
            if left >= right:
                return

            pivot_idx = partition(left, right)
            if pivot_idx == k:
                return
            elif pivot_idx < k:
                quick_select(pivot_idx + 1, right, k)
            else:
                quick_select(left, pivot_idx - 1, k)

        quick_select(0, len(items) - 1, k - 1)
        return [num for num, _ in items[:k]]
